### ISSUE TITLE
data loader should record failed imports

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#255]
_Imported from trac ticket 255.
Created by **rneville**
Opened in trac: Aug 11, 2010, 11:06:01 AM
Last modified in trac: May 3, 2011, 10:07:49 PM_
***
_by rneville:_
loading of system/mock data shows a # of failed records, it should write a file somewhere that shows which records failed and the SQL/SQLa error code that was thrown.

**Comment by kentbower** [comment]
_**kbower** trac comment on Aug 11, 2010, 6:11:03 PM:_

Yes, it is a pain to turn on engine.echo to figure out what is going on, I know.  What is annoying is the lack of error distinction between primary key duplication and one of many other "IntegrityErrors"... they all are the same python Exception.

**Comment by kentbower** [comment]
_**kbower** trac comment on Aug 12, 2010, 2:00:54 PM:_

I've got reservations about implementing this.  If we are loading many rows (like thousands, then we _expect_ those to fail on a later upgrade).  My concern is the additional time writing log files if this is already going to be slow as it is for thousands of rows.

What do you think?

Is there a way around needing to load 42,000 zip codes?  That table maybe only needs be loaded one time?

**Comment by kentbower** [comment]
_**rneville** trac comment on Aug 12, 2010, 7:49:35 PM:_

agreed that it only needs to be loaded once...but it highlights the impetus behind my opening of this ticket. I have a table that I expect to load completely, without error...there are 10's of thousands of rows, how do I find the needle in the haystack? I was thinking of something along the lines of the way DCS's data_load scripts used to work in that a file of failed records would be created every time or something like that and if they already existed, just overwrite them.

**Comment by kentbower** [comment]
_**kbower** trac comment on Aug 12, 2010, 8:36:56 PM:_

yeah, I was thinking something along those lines...

are you waiting on this?  (I didn't think so because you marked it 'minor' and also because I showed you in the email where to uncomment the print statement.  If you do droptables and then uncomment the print you will see exactly what is going on... either have duplicates, missing field in non-null, or, my guess is a field too many characters).

Let me know if you are waiting on this, I can make it a higher priority.

**Comment by kentbower** [comment]
_**rneville** trac comment on Aug 13, 2010, 8:22:52 AM:_

no this isn't a show stopper...

**Comment by kentbower** [comment]
_**kbower** trac comment on Oct 20, 2010, 1:20:15 PM:_

For now, I think we should just uncomment the `#print e` statement at the end of dataload.py

```python
def add_to_database(obj):
    """
    Add to the database unless already there
    """
    DBSession.begin_nested() # savepoint
    DBSession.add(obj)
    try:
        DBSession.flush()
    except IntegrityError as e:
        ret = False
        #print e #uncomment me to see errors
        DBSession.rollback()
    else:
        ret = True
        # since this is a nested transaction, this is not a true
        # commit, but rather releases the savepoint
        DBSession.commit()
    return ret
```
My concern is that writing all errors will be way too much since there are thousands and thousands that are expected errors.

**Comment by kentbower** [comment]
_**kbower** trac status change on Oct 20, 2010, 1:20:15 PM:_ **closed**

**Comment by kentbower** [comment]
_**kbower** trac comment on May 3, 2011, 10:07:49 PM:_

(Changeset [[1428]](../commit/79ecc85e92f256476b0535eff56535722133673e))
 * fix circular referencing foreign keys with use_alter and deferrable
 * make most warnings raise error as recommended by Mike Bayer and me
 * fixes #255 for dataload, errors should be thrown if they are referential or constraint error or warning if tables just being added
 * fix the weberror extra exception that is always raised
 * fix some foreign keys, many were defined backwards in the mappers
